src/basset.cpp

```diff
@@ -190,79 +160,107 @@ void Regex::report(int errcode) const {
 }
 
-void CompilationDatabase::add(const string &directory,
-                              const vector<string> &command) {
-  vector<string> files;
+bool CompilationDatabase::load() {
+  ifstream ifs(filename);
+  if (!ifs) {
+    return true;
+  }
 
-  for (auto &&argument : command) {
-    if (is_source_file(argument)) {
-      files.emplace_back(argument);
+  try {
+    auto json = nlohmann::json::parse(ifs);
+
+    // load existing database with removing entries for no longer existing
+    // sources
+    // (https://github.com/rizsotto/Bear/wiki/Features#append-to-existing-jspn-cdb)
+    for (auto it = json.begin(); it != json.end(); ++it) {
+      const auto directory = (*it)["directory"].get<string>();
+      const auto file = (*it)["file"].get<string>();
+
+      if (exists(path(directory) / file)) {
+        index[make_index_key(directory, file)] =
+            make_pair(directory, (*it)["arguments"]);
+      }
     }
+  } catch (const nlohmann::json::parse_error &e) {
+    cerr << "json error: " << e.what() << '\n';
+    return false;
   }
 
-  for (const auto &file : files) {
-    if (first) {
-      first = false;
-    } else {
-      *this << ',';
-    }
+  return true;
+}
 
-    *this << "\n"
-             "  {\n"
-             // clang-format off
-             "    \"directory\": \"" << json_escape(directory) << "\",\n"
-             // clang-format on
-             "    \"arguments\": [";
+bool CompilationDatabase::save() {
+  ofstream ofs(filename);
+  if (!ofs) {
+    return false;
+  }
 
-    bool first_arg{true};
+  auto json = nlohmann::json::array();
+  for (const auto &kv : index) {
+    json += nlohmann::json::object({{"directory", kv.second.first},
+                                    {"file", kv.first},
+                                    {"arguments", kv.second.second}});
+  }
 
-    for (const auto &arg : command) {
-      if (first_arg) {
-        first_arg = false;
-      } else {
-        *this << ',';
-      }
+  ofs << json.dump(4) << '\n';
+  return true;
+}
 
-      *this << "\n"
-               // clang-format off
-               "      \"" << json_escape(arg) << '\"';
-      // clang-format on
+void CompilationDatabase::add(const string &directory,
+                              const vector<string> &command) {
+  for (auto &&argument : command) {
+    if (!is_source_file(argument)) {
+      continue;
     }
 
-    *this << "\n"
-             "    ],\n"
-             // clang-format off
-             "    \"file\": \"" << json_escape(file) << "\"\n"
-             // clang-format on
-             "  }";
+    // append new entry for (directory, filename) or update it if it is already
+    // existing
+    const auto &it = index.find(make_index_key(directory, argument));
+    if (it != index.end()) {
+      it->second = make_pair(directory, command);
+    } else {
+      index[make_index_key(directory, argument)] =
+          make_pair(directory, command);
+    }
```

Comment:
I suppose you can shorten this code using [insert_or_assign()](https://en.cppreference.com/w/cpp/container/unordered_map/insert_or_assign).